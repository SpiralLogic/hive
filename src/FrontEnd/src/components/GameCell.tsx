import { FunctionComponent, h } from 'preact';
import { HexCoordinates, Tile as TileType } from '../domain';
import { MoveEvent, TileEvent } from '../utilities/hive-dispatcher';
import { addHiveDispatchListener, dispatchHiveEvent, useClassReducer } from '../utilities/hooks';
import { handleDragOver, handleKeyboardNav, isEnterOrSpace } from '../utilities/handlers';
import { useEffect, useState } from 'preact/hooks';
import Hexagon from './Hexagon';

type Props = { coords: HexCoordinates; hidden?: boolean };
const GameCell: FunctionComponent<Props> = (props) => {
  const { coords, children, hidden } = props;
  const [classes, setClasses] = useClassReducer('hide');
  const [selectedTile, setSelectedTile] = useState<TileType | null>(null);

  const isValidMove = (validMoves: HexCoordinates[]) =>
    validMoves.some((dest) => dest.q == coords.q && dest.r == coords.r);
  useEffect(() => setClasses({ type: hidden ? 'add' : 'remove', classes: ['hide'] }), [hidden]);

  addHiveDispatchListener<TileEvent>('tileDeselected', (event) => {
    if (!isValidMove(event.tile.moves)) {
      setClasses({ type: 'remove', classes: ['no-drop'] });
    }
    setSelectedTile(null);
    setClasses({ type: 'remove', classes: ['active', 'can-drop'] });
  });

  addHiveDispatchListener<TileEvent>('tileSelected', (event) => {
    if (!isValidMove(event.tile.moves)) {
      setClasses({ type: 'add', classes: ['no-drop'] });
    } else {
      setSelectedTile(event.tile);
      setClasses({ type: 'add', classes: ['can-drop'] });
    }
  });

  addHiveDispatchListener<TileEvent>('tileDropped', () => {
    if (classes.includes('active') && selectedTile && isValidMove(selectedTile.moves)) {
      dispatchHiveEvent<MoveEvent>({
        type: 'move',
        move: { coords, tileId: selectedTile.id },
      });
    }
  });

  const handleDragLeave = (event: DragEvent) => {
    event.stopPropagation();
    setClasses({ type: 'remove', classes: ['active'] });
  };

  const handleDragEnter = () => {
    setClasses({ type: 'add', classes: ['active'] });
  };

  const handleClick = (event: UIEvent) => {
    if (!(selectedTile && isValidMove(selectedTile.moves))) return;
    event.stopPropagation();
    dispatchHiveEvent({ type: 'move', move: { coords, tileId: selectedTile.id } });
    dispatchHiveEvent({ type: 'tileClear', tile: selectedTile });
  };

  const handleKeydown = (e: KeyboardEvent) => {
    if (!handleKeyboardNav(e) && isEnterOrSpace(e)) return handleClick(e);
  };

  const attributes = {
    class: classes || undefined,
    tabindex: selectedTile && isValidMove(selectedTile.moves) ? 0 : undefined,
    role: hidden ? 'none' : 'cell',
  };

  const handlers = {
    ondragover: handleDragOver,
    ondragleave: handleDragLeave,
    ondragenter: handleDragEnter,
    onmouseenter: handleDragEnter,
    onmouseleave: handleDragLeave,
    onclick: handleClick,
    onkeydown: handleKeydown,
  };
  return (
    <Hexagon hidden={hidden} {...attributes} {...handlers}>
      {children}
    </Hexagon>
  );
};
GameCell.displayName = 'Cell';
export default GameCell;
