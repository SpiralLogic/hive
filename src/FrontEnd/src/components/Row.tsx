import { deepEqual } from 'fast-equals';
import { FunctionComponent, h } from 'preact';
import { memo } from 'preact/compat';
import { HexCoordinates } from '../domain';
import Cell from './Cell';

type CellProps = typeof Cell.arguments['props'];
const cellKey = ({ q, r }: HexCoordinates) => `${q}-${r}`;
const createCell = (cell: CellProps) => <Cell key={cellKey(cell.coords)} {...cell} />;
const createPlaceholder = (key: number) => <div key={key} className="hidden" />;

type Props = {
    id: number;
    row: Array<CellProps>;
};

const Row: FunctionComponent<Props> = (props: Props) => {
    const { row } = props;
    const cells = row.map((cell, i) => (cell && createCell(cell)) || createPlaceholder(i));

    return <div className="hex-row">{cells}</div>;
};

Row.displayName = 'Row';
export default memo(Row, deepEqual);
