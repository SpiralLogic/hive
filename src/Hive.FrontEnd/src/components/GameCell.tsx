import { FunctionComponent } from 'preact';
import { useEffect, useState } from 'preact/hooks';
import { HexCoordinates, Tile as TileType } from '../domain';
import { TileEvent } from '../services';
import { addHiveDispatchListener, dispatchHiveEvent } from '../utilities/dispatcher';
import { handleDragOver, handleKeyboardNav, isEnterOrSpace } from '../utilities/handlers';
import { useClassReducer } from '../utilities/class-reducer';
import Hexagon from './Hexagon';
const isValidMove = (validMoves: HexCoordinates[], coords: HexCoordinates) =>
  validMoves.some((dest) => dest.q == coords.q && dest.r == coords.r);

type Props = { coords: HexCoordinates; hidden?: boolean };
const GameCell: FunctionComponent<Props> = (props) => {
  const { coords, children, hidden } = props;
  const [classes, setClasses] = useClassReducer('hide');
  const [selectedTile, setSelectedTile] = useState<TileType | null>(null);

  useEffect(() => setClasses({ type: hidden ? 'add' : 'remove', classes: ['hide'] }), [hidden]);

  addHiveDispatchListener<TileEvent>('tileDeselected', () => {
    setSelectedTile(null);
    setClasses({ type: 'remove', classes: ['active', 'can-drop', 'no-drop'] });
  });

  addHiveDispatchListener<TileEvent>('tileSelected', (intent) => {
    if (!isValidMove(intent.tile.moves, coords)) {
      setClasses({ type: 'add', classes: ['no-drop'] });
    } else {
      setSelectedTile(intent.tile);
      setClasses({ type: 'add', classes: ['can-drop'] });
    }
  });

  addHiveDispatchListener<TileEvent>('tileDropped', () => {
    if (classes.includes('active') && selectedTile && isValidMove(selectedTile.moves, coords)) {
      dispatchHiveEvent({
        type: 'move',
        move: { coords, tileId: selectedTile.id },
      });
    }
  });

  const handleDragLeave = (event: DragEvent) => {
    event.stopPropagation();
    setClasses({ type: 'remove', classes: ['active'] });
  };

  const handleDragEnter = () => {
    setClasses({ type: 'add', classes: ['active'] });
  };

  const handleClick = (event: UIEvent) => {
    if (!(selectedTile && isValidMove(selectedTile.moves, coords))) return;
    event.stopPropagation();
    dispatchHiveEvent({ type: 'tileClear', tile: selectedTile });
    dispatchHiveEvent({ type: 'move', move: { coords, tileId: selectedTile.id } });
  };

  const handleKeydown = (e: KeyboardEvent) => {
    if (handleKeyboardNav(e) || !isEnterOrSpace(e)) return;
    handleClick(e);
  };

  const attributes = {
    class: classes || undefined,
    tabindex: selectedTile && isValidMove(selectedTile.moves, coords) ? 0 : undefined,
    role: hidden ? 'none' : 'cell',
  };

  const handlers = {
    ondragover: handleDragOver,
    ondragleave: handleDragLeave,
    ondragenter: handleDragEnter,
    onmouseenter: handleDragEnter,
    onmouseleave: handleDragLeave,
    onclick: handleClick,
    onkeydown: handleKeydown,
  };
  return (
    <Hexagon hidden={hidden} {...attributes} {...handlers}>
      {children}
    </Hexagon>
  );
};
GameCell.displayName = 'Cell';
export default GameCell;
