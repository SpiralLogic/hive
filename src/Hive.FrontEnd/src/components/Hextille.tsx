import '../css/hextille.css';

import { FunctionComponent, toChildArray, VNode } from 'preact';

import { Row } from '../domain';

const Hextille: FunctionComponent<{ class?: string }> = (props) => {
  const childArray = toChildArray(props.children);
  const shiftClass = ((childArray[0] as VNode<Row>).key ?? childArray.length) % 2 ? 'left' : 'right';

  return (
    <div role="grid" class={shiftClass}>
      {props.children}
    </div>
  );
};

Hextille.displayName = 'Hextille';
export default Hextille;
