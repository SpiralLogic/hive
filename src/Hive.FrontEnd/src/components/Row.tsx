import '../css/row.css';

import { FunctionComponent } from 'preact';

type Props = { class?: string; zIndex?: number; hidden?: boolean };

const Row: FunctionComponent<Props> = (props) => {
  const { hidden, zIndex } = props;
  return (
    <div class={props.class} role={hidden ? 'none' : 'row'} style={zIndex ? { zIndex } : undefined}>
      {props.children}
    </div>
  );
};

Row.displayName = 'Row';
export default Row;
