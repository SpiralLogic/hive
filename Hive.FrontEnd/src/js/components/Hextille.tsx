import * as React from 'react';
import {Hexagon} from '../domain';
import Row from './Row';

const getWidth = (hexagons: Hexagon[]) => {
    const [min, max] = hexagons.reduce(([min, max], c) =>
        [Math.min(min, c.coordinates.q), Math.max(max, c.coordinates.q)], [0, 0]);
    return [min, max - min + 1];
};

function getHeight(sortedHexagons: Hexagon[]) {
    const firstRow = sortedHexagons[0].coordinates.r;
    const height = sortedHexagons[sortedHexagons.length - 1].coordinates.r - firstRow + 1;
    return [firstRow, height];
}


function createRows(sortedHexagons: Hexagon[]) {
    const [firstRow, height] = getHeight(sortedHexagons);
    const [firstColumn, width] = getWidth(sortedHexagons);

    const createEmptyRows = () => {
        const createEmptyRow = (i: number) => ({
            r: firstRow + i,
            row: new Array(width).fill(false),
            firstColumn
        });
        return Array.from(Array(height).keys(), createEmptyRow);
    };

    return sortedHexagons.reduce((rows, hexagon) => {
        rows[hexagon.coordinates.r - firstRow].row[hexagon.coordinates.q - firstColumn] = hexagon;
        return rows;
    }, createEmptyRows());
}

export const Hextille: React.FunctionComponent<{ hexagons: Hexagon[] }> = ({hexagons}) => {

    if (!hexagons.length) throw Error('Nothing to render !');
    const sortedHexagons = hexagons.sort((c1, c2) => c1.coordinates.r - c2.coordinates.r || c1.coordinates.q - c2.coordinates.q);
    const shiftClass = (sortedHexagons[0].coordinates.r % 2) ? 'left' : 'right';
    const rows = createRows(sortedHexagons);

    return (
        <div className={'hextille ' + shiftClass}>
            {rows.map((row) => (
                <Row key={row.r} {...row} />
            ))}
        </div>
    );
};

Hextille.displayName = 'Hextille';