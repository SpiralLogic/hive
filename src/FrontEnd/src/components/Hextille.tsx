import { Cell } from '../domain';
import { FunctionComponent, h } from 'preact';
import Row from './Row';

type Row = { id: number; row: Cell[] };

function getWidth (cells: Cell[]): [number, number] {
    const [min, max] = cells.reduce(
        ([min, max], c) => [Math.min(min, c.coords.q), Math.max(max, c.coords.q)],
        [0, 0]
    );
    return [min, max - min + 1];
}

function getHeight (sortedHexagons: Cell[]): [number, number] {
    const firstCell = sortedHexagons[0] as Cell;
    const lastCell = sortedHexagons[sortedHexagons.length - 1] as Cell;
    const height = lastCell.coords.r - firstCell.coords.r + 1;
    return [firstCell.coords.r, height];
}

function createRows (sortedHexagons: Cell[]) {
    const [firstRow, height] = getHeight(sortedHexagons);
    const [firstColumn, width] = getWidth(sortedHexagons);

    const createEmptyRow = (i: number): Row => ({
        id: firstRow + i,
        row: new Array(width).fill(false)
    });

    const createEmptyRows = () => {
        return Array.from(Array(height).keys(), createEmptyRow);
    };

    return sortedHexagons.reduce((rows, cell) => {
        (rows[cell.coords.r - firstRow] as Row).row[cell.coords.q - firstColumn] = cell;
        return rows;
    }, createEmptyRows());
}

type Props = { cells: Cell[] };

const Hextille: FunctionComponent<Props> = (props: Props) => {
    const { cells } = props;
    const sortedHexagons = cells.sort(
        (c1, c2) => c1.coords.r - c2.coords.r || c1.coords.q - c2.coords.q
    );
    const shiftClass = sortedHexagons[0].coords.r % 2 ? 'left' : 'right';
    const rows = createRows(sortedHexagons);

    return (
        <div className="hex-container">
            <div className={'hextille ' + shiftClass}>
                {rows.map((row) => (
                    <Row key={row.id} {...row} />
                ))}
            </div>

        </div>
    );
};

Hextille.displayName = 'Hextille';
export default Hextille;
