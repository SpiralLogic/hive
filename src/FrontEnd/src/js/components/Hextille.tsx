import { FunctionComponent, h } from 'preact';
import { Cell } from '../domain';
import Row from './Row';

function getWidth(cells: Cell[]) {
    const [min, max] = cells.reduce(([min, max], c) => [Math.min(min, c.coords.q), Math.max(max, c.coords.q)], [0, 0]);
    return [min, max - min + 1];
}

function getHeight(sortedHexagons: Cell[]) {
    const firstRow = sortedHexagons[0].coords.r;
    const height = sortedHexagons[sortedHexagons.length - 1].coords.r - firstRow + 1;
    return [firstRow, height];
}

function createRows(sortedHexagons: Cell[]) {
    const [firstRow, height] = getHeight(sortedHexagons);
    const [firstColumn, width] = getWidth(sortedHexagons);

    const createEmptyRow = (i: number) => ({
        id: firstRow + i,
        row: new Array(width).fill(false),
    });

    const createEmptyRows = () => {
        return Array.from(Array(height).keys(), createEmptyRow);
    };

    return sortedHexagons.reduce((rows, cell) => {
        rows[cell.coords.r - firstRow].row[cell.coords.q - firstColumn] = cell;
        return rows;
    }, createEmptyRows());
}

type Props = { cells: Cell[] };

const Hextille: FunctionComponent<Props> = (props: Props) => {
    const { cells } = props;
    const sortedHexagons = cells.sort((c1, c2) => c1.coords.r - c2.coords.r || c1.coords.q - c2.coords.q);
    const shiftClass = sortedHexagons[0].coords.r % 2 ? 'left' : 'right';
    const rows = createRows(sortedHexagons);

    return (
        <div className={'hextille ' + shiftClass}>
            {rows.map((row) => (
                <Row key={row.id} {...row} />
            ))}
        </div>
    );
};

Hextille.displayName = 'Hextille';
export default Hextille;
