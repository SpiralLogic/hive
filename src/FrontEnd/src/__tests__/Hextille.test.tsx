import { h } from 'preact';
import GameCell from '../components/GameCell';
import Hextille from '../components/Hextille';
import Row from '../components/Row';
import { renderElement } from './test-helpers';

describe('hextille Tests', () => {
  it('matches current snapshot', () => {
    expect(
      renderElement(
        <Hextille>
          <Row>
            <GameCell coords={{ q: 0, r: 0 }} />
          </Row>
        </Hextille>
      )
    ).toMatchSnapshot();
  });
});
