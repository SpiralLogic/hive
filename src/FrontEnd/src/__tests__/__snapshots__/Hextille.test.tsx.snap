// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hextille Tests matches current snapshot 1`] = `
<div
  class="left"
  role="grid"
>
  <div
    role="row"
  >
    <div
      role="cell"
    >
      <svg
        viewBox="0 0 100 100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          href="#hex"
        />
      </svg>
    </div>
  </div>
</div>
`;
